File: examples/go/cases/tasks/task_list/task_list.api.case.go
package task_list

import (
	"encoding/json"
	"fmt"

	"app-protocol/examples/go/core"
	"app-protocol/examples/go/core/shared"
)

type TaskStore interface {
	Read() ([]map[string]any, error)
	Write(value []map[string]any) error
	Reset() error
}

type API struct {
	Ctx        *core.ApiContext
	domainCase *Domain
}

func NewAPI(ctx *core.ApiContext) *API {
	return &API{
		Ctx:        ctx,
		domainCase: NewDomain(),
	}
}

func (c *API) Handler(input TaskListInput) core.ApiResponse[TaskListOutput] {
	result := core.ExecuteAPI(
		input,
		c.validate,
		nil,
		nil,
		c.service,
	)

	if result.Success {
		result.StatusCode = 200
		return result
	}

	result.StatusCode = mapErrorCodeToStatus(result.Error)
	return result
}

func (c *API) HandleAny(input any) (any, error) {
	parsed, err := c.domainCase.ParseInput(input)
	if err != nil {
		return c.failure("VALIDATION_FAILED", err, 400), nil
	}

	return c.Handler(parsed), nil
}

func (c *API) Router() core.RouteBinding {
	return core.RouteBinding{
		Method: "GET",
		Path:   "/tasks",
		Handler: func(_ core.RouteRequest) (any, error) {
			return c.Handler(TaskListInput{}), nil
		},
	}
}

func (c *API) Test() error {
	store, err := c.resolveTaskStore()
	if err != nil {
		return err
	}

	if err := store.Reset(); err != nil {
		return err
	}

	if err := store.Write([]map[string]any{
		mustTaskRecord(Task{
			ID:        "task_001",
			Title:     "Older task",
			Status:    TaskStatusTodo,
			CreatedAt: "2026-03-18T12:00:00.000Z",
			UpdatedAt: "2026-03-18T12:00:00.000Z",
		}),
		mustTaskRecord(Task{
			ID:        "task_002",
			Title:     "Newer task",
			Status:    TaskStatusDoing,
			CreatedAt: "2026-03-18T12:30:00.000Z",
			UpdatedAt: "2026-03-18T12:30:00.000Z",
		}),
	}); err != nil {
		return err
	}

	result := c.Handler(TaskListInput{})
	if !result.Success || result.Data == nil {
		return fmt.Errorf("test: handler should return a successful task list")
	}

	if len(result.Data.Tasks) != 2 {
		return fmt.Errorf("test: task list should return all persisted tasks")
	}

	if result.Data.Tasks[0].ID != "task_002" {
		return fmt.Errorf("test: task list should sort by createdAt descending")
	}

	if err := store.Write([]map[string]any{
		{
			"id":        "broken",
			"title":     "Broken task",
			"status":    "broken",
			"createdAt": "2026-03-18T12:00:00.000Z",
			"updatedAt": "2026-03-18T12:00:00.000Z",
		},
	}); err != nil {
		return err
	}

	invalidResult := c.Handler(TaskListInput{})
	if invalidResult.Success {
		return fmt.Errorf("test: invalid persisted records must return failure")
	}

	return nil
}

func (c *API) validate(input TaskListInput) error {
	if err := c.domainCase.Validate(input); err != nil {
		return &shared.AppCaseError{
			Code:    "VALIDATION_FAILED",
			Message: err.Error(),
		}
	}

	return nil
}

func (c *API) service(_ TaskListInput) (TaskListOutput, error) {
	taskStore, err := c.resolveTaskStore()
	if err != nil {
		return TaskListOutput{}, err
	}

	records, err := taskStore.Read()
	if err != nil {
		return TaskListOutput{}, err
	}

	tasks := make([]Task, 0, len(records))
	for index, record := range records {
		task, err := decodeTask(record)
		if err != nil {
			return TaskListOutput{}, &shared.AppCaseError{
				Code:    "INTERNAL",
				Message: fmt.Sprintf("task_list.persisted_tasks[%d]: %s", index, err.Error()),
			}
		}

		if err := c.domainCase.ValidateTask(task); err != nil {
			return TaskListOutput{}, &shared.AppCaseError{
				Code:    "INTERNAL",
				Message: err.Error(),
			}
		}

		tasks = append(tasks, task)
	}

	output := TaskListOutput{
		Tasks: c.domainCase.SortTasks(tasks),
	}
	if err := c.domainCase.ValidateOutput(output); err != nil {
		return TaskListOutput{}, err
	}

	return output, nil
}

func (c *API) resolveTaskStore() (TaskStore, error) {
	if c.Ctx == nil || c.Ctx.Extra == nil {
		return nil, &shared.AppCaseError{Code: "INTERNAL", Message: "task_list requires a configured taskStore provider"}
	}

	providers, ok := c.Ctx.Extra["providers"].(shared.Dict)
	if !ok {
		return nil, &shared.AppCaseError{Code: "INTERNAL", Message: "task_list requires a configured taskStore provider"}
	}

	taskStore, ok := providers["taskStore"].(TaskStore)
	if !ok || taskStore == nil {
		return nil, &shared.AppCaseError{Code: "INTERNAL", Message: "task_list requires a configured taskStore provider"}
	}

	return taskStore, nil
}

func (c *API) failure(code string, err error, statusCode int) core.ApiResponse[TaskListOutput] {
	appErr := shared.ToAppError(&shared.AppCaseError{
		Code:    code,
		Message: err.Error(),
	}, code)

	return core.ApiResponse[TaskListOutput]{
		AppResult: shared.AppResult[TaskListOutput]{
			Success: false,
			Error:   &appErr,
		},
		StatusCode: statusCode,
	}
}

func mapErrorCodeToStatus(err *shared.AppError) int {
	if err == nil {
		return 500
	}

	switch err.Code {
	case "VALIDATION_FAILED":
		return 400
	default:
		return 500
	}
}

func decodeTask(value any) (Task, error) {
	var task Task
	encoded, err := json.Marshal(value)
	if err != nil {
		return task, err
	}

	if err := json.Unmarshal(encoded, &task); err != nil {
		return task, err
	}

	return task, nil
}

func mustTaskRecord(task Task) map[string]any {
	record, _ := json.Marshal(task)
	decoded := map[string]any{}
	_ = json.Unmarshal(record, &decoded)
	return decoded
}

File: examples/python/cases/tasks/task_list/task_list_api_case.py
# ==========================================================================
# APP v1.1.4
# cases/tasks/task_list/task_list_api_case.py
# --------------------------------------------------------------------------
# API surface for the task_list Case.
#
# Responsibility:
# - expose the list-tasks capability via GET /tasks
# - delegate input validation to the domain surface
# - read from task_store, assert output integrity, sort by createdAt desc
# - return a structured ApiResponse (never throw on data corruption)
#
# Execution center: _service (atomic — no cross-case composition)
# ==========================================================================

from __future__ import annotations

import uuid
from typing import Any, Literal

from core.api_case import ApiContext, ApiResponse, BaseApiCase
from core.shared.app_structural_contracts import AppCaseError

from cases.tasks.task_list.task_list_domain_case import (
    VALID_STATUSES,
    TaskListDomain,
    assert_task_collection,
)

# ==========================================================================
# Shared types
# --------------------------------------------------------------------------
# Cases are self-contained — types are redefined per file.
# ==========================================================================

TaskStatus = Literal["todo", "doing", "done"]


class Task(dict):  # type: ignore[type-arg]
    """Runtime-typed task record (id, title, status, createdAt, updatedAt)."""


class TaskListInput(dict):  # type: ignore[type-arg]
    """Empty input — no filters accepted."""


class TaskListOutput(dict):  # type: ignore[type-arg]
    """Output containing a list of Task records."""


# ==========================================================================
# TaskListApi
# ==========================================================================

class TaskListApi(BaseApiCase[TaskListInput, TaskListOutput]):
    """API surface for the task_list Case."""

    # ==================================================================
    # Required: handler and router
    # ==================================================================

    def handler(self, input: TaskListInput) -> ApiResponse[TaskListOutput]:
        """
        Main capability handler.

        Delegates to the standard execute() pipeline which calls
        _validate -> _authorize -> _service in sequence.
        Returns 200 on success, or a structured failure response on error.
        """
        response = self.execute(input)
        if response.success:
            return ApiResponse(success=True, data=response.data, status_code=200)
        return response

    def router(self) -> dict[str, Any]:
        """
        Transport binding for GET /tasks.

        The host/adapter collects router() from each Case to mount routes.
        This method never contains business logic — it only maps the transport.
        """
        return {
            "method": "GET",
            "path": "/tasks",
            "handler": self.handler,
        }

    # ==================================================================
    # Protected hooks
    # ==================================================================

    def _validate(self, input: TaskListInput) -> None:
        """
        Delegate input validation to the domain surface.

        Instantiates a fresh TaskListDomain and calls validate().
        Raises AppCaseError("VALIDATION_FAILED", ...) on invalid input.
        """
        domain = TaskListDomain()
        domain.validate(input)

    def _service(self, input: TaskListInput) -> TaskListOutput:
        """
        Main execution logic (atomic Case).

        Steps:
        1. Read all tasks from ctx.extra["task_store"].
        2. Assert structural integrity via assert_task_collection from domain.
        3. Sort tasks by createdAt descending (newest first).
        4. Return TaskListOutput.

        Corrupted data raises AppCaseError("INTERNAL", ...) which the base
        execute() pipeline catches and wraps into ApiResponse(success=False).
        """
        task_store = self.ctx["extra"]["task_store"]
        raw_tasks: Any = task_store.read()

        assert_task_collection(raw_tasks)

        # Sort by createdAt descending — newest first.
        sorted_tasks = sorted(
            raw_tasks,
            key=lambda t: t.get("createdAt", ""),
            reverse=True,
        )

        return {"tasks": sorted_tasks}  # type: ignore[return-value]

    # ==================================================================
    # Test
    # ==================================================================

    def test(self) -> None:
        """
        Self-contained API surface test.

        Covers:
        - empty store returns empty tasks list
        - store with 2 tasks returns them in createdAt descending order
        - corrupted data returns success=False with INTERNAL error (no throw)
        """
        class _InMemoryTaskStore:
            def __init__(self, initial_data: list[dict[str, Any]]) -> None:
                self._data = list(initial_data)

            def read(self) -> list[dict[str, Any]]:
                return list(self._data)

            def write(self, tasks: list[dict[str, Any]]) -> None:
                self._data = list(tasks)

        class _NoopLogger:
            def debug(self, msg: str, meta: Any = None) -> None: pass
            def info(self, msg: str, meta: Any = None) -> None: pass
            def warn(self, msg: str, meta: Any = None) -> None: pass
            def error(self, msg: str, meta: Any = None) -> None: pass

        # ----------------------------------------------------------------
        # Scenario 1: empty store returns empty tasks
        # ----------------------------------------------------------------
        store = _InMemoryTaskStore([])
        ctx: ApiContext = ApiContext(
            correlation_id=str(uuid.uuid4()),
            execution_id=str(uuid.uuid4()),
            logger=_NoopLogger(),
            extra={"task_store": store},
        )
        api = TaskListApi(ctx)

        response = api.handler({})
        assert response.success is True, "Empty store: response must be success"
        assert response.data is not None, "Empty store: data must not be None"
        assert response.data["tasks"] == [], (
            "Empty store: tasks must be an empty list"
        )
        assert response.status_code == 200, "Success response must be 200"

        # ----------------------------------------------------------------
        # Scenario 2: store with 2 tasks — result ordered newest first
        # ----------------------------------------------------------------
        task_older = {
            "id": "task-1",
            "title": "Older task",
            "status": "todo",
            "createdAt": "2024-01-01T08:00:00Z",
            "updatedAt": "2024-01-01T08:00:00Z",
        }
        task_newer = {
            "id": "task-2",
            "title": "Newer task",
            "status": "doing",
            "createdAt": "2024-01-02T10:00:00Z",
            "updatedAt": "2024-01-02T10:00:00Z",
        }
        store.write([task_older, task_newer])

        response2 = api.handler({})
        assert response2.success is True, "Fixture store: response must be success"
        tasks = response2.data["tasks"]  # type: ignore[index]
        assert len(tasks) == 2, "Must return exactly 2 tasks"
        assert tasks[0]["id"] == "task-2", (
            "Newest task (task-2) must appear first"
        )
        assert tasks[1]["id"] == "task-1", (
            "Older task (task-1) must appear second"
        )

        # ----------------------------------------------------------------
        # Scenario 3: corrupted data returns structured failure — no throw
        # ----------------------------------------------------------------
        store_corrupt = _InMemoryTaskStore([{"id": "bad-task"}])  # type: ignore[list-item]

        ctx_corrupt: ApiContext = ApiContext(
            correlation_id=str(uuid.uuid4()),
            execution_id=str(uuid.uuid4()),
            logger=_NoopLogger(),
            extra={"task_store": store_corrupt},
        )
        api_corrupt = TaskListApi(ctx_corrupt)

        response3 = api_corrupt.handler({})
        assert response3.success is False, (
            "Corrupted data: response must be failure (success=False)"
        )
        assert response3.error is not None, (
            "Corrupted data: error must be present"
        )
        assert response3.error.code == "INTERNAL", (
            f"Corrupted data: expected INTERNAL error, got {response3.error.code}"
        )
